fix: kube-core::Schema hoisting logic

NickLarsenNZ opened the pull request:
Fixes https://github.com/kube-rs/kube/issues/1821

This updates the kube Schema hoisting since schemars Schema's have changed.

- Add tests for a variety of enum use-cases:
  - Tagged vs Untagged
  - Unit vs Tuple vs Structural variants
  - With and without doc-comments (descriptions)
- Rewrite the hoisting logic
  - This is annotated with dev-comments to help understand intend and to ease future schemars changes.

This also fixes other issues that appeared along the way:
- Untagged enum variant doc-comments were being applied to field descriptions.
- Additional `null` entry added to enums.

See also: https://github.com/stackabletech/kube-rs/pull/1, where this work was originally done.

---------

<!--
Thank you for your Pull Request. Please provide a description above and review the requirements below.

Bug fixes and new features should include tests.

Contributors guide: https://github.com/kube-rs/kube-rs/blob/master/CONTRIBUTING.md
-->

## Motivation

Since kube 1.0.0, schemas became invalid when optional enums contained comments on the variants.

## Solution

- Break out the hoisting logic into separate modules with scoped tests.
  - Heavily document the steps to help developers with future schemars Schema generation changes.
- Add tests to try and catch this in future.

sbernauer committed:
fix: kube-core::Schema hoisting logic (#1)

Fixes kube Schema conversion since schemars Schema's have changed.

- Add tests for a variety of enum use-cases:
  - Tagged vs Untagged
  - Unit vs Tuple vs Structural variants
  - With and without doc-comments (descriptions)
- Rewrite the hoisting logic
  - This is annotated with dev-comments to help understand intend and to ease future schemars changes.

This also fixes other issues:
- Untagged enum variant doc-comments were being applied to field descriptions.
- Additional `null` entry added to enums.

---------

Co-authored-by: Nick Larsen <nick.larsen@stackable.tech>
Co-authored-by: Sebastian Bernauer <sebastian.bernauer@stackable.tech>
Co-authored-by: Techassi <git@techassi.dev>
Signed-off-by: Nick Larsen <nick.larsen@stackable.tech>

NickLarsenNZ commented:
I'm unsure what to do about the CI error:

```
error: rustc 1.86.0 is not supported by the following package:
  home@0.5.12 requires rustc 1.88
```

For me, I get home@0.5.9. Should I pin the dependency?
https://github.com/kube-rs/kube/blob/7c63f56828532b2ee98002a98d51d4933ccf42f7/Cargo.toml#L50

Techassi committed:
chore: Merge branch 'main' into fix/schema-hoisting

Techassi commented:
I've just synced this PR branch to include the CI fixes (done in #1840). This PR should now be ready for review.

NickLarsenNZ commented:
Hi @clux, are you able to take a look at this?

We saw a new PR come up that fixes part of the problem: https://github.com/kube-rs/kube/pull/1853

NickLarsenNZ committed:
chore(schema): Split StructuralSchemaRewriter into multiple transformers

Signed-off-by: Nick Larsen <nick.larsen@stackable.tech>

NickLarsenNZ commented:
I have updated the implementation to split the single transformer into multiple transformers.

The original `StructuralSchemaRewriter` should be renamed because it now does two random other things (should it just be nameds `OtherSchemaRewriter`, or should they too be split into separate Transform impls?

I did try moving the body of the transform functions into the Transform impl but then the unit tests would need rewriting (I can try that is you prefer).

Anyway, I will leave it as is and await your feedback.

NickLarsenNZ committed:
Merge branch 'main' into fix/schema-hoisting

clux commented on kube-core/src/schema/transform_one_of.rs:
```diff
@@ -0,0 +1,181 @@
+///
+/// This will return early without modifications unless:
+/// - There are `oneOf` subschemas (not empty).
+/// - Each subschema contains an `enum` or `const`.
```
Possibly this can be reduced in scope if we use the existing [ReplaceConstValue](https://docs.rs/schemars/latest/schemars/transform/struct.ReplaceConstValue.html)? it [doesn't do a lot](https://docs.rs/schemars/latest/src/schemars/transform.rs.html#369) though.

clux commented on kube-core/src/schema/transform_any_of.rs:
```diff
@@ -0,0 +1,238 @@
+
+    if to_hoist.metadata.is_none() {
+        to_hoist.metadata = kube_schema_metadata;
+    }
```
```suggestion
    let mut to_hoist = to_hoist.clone();
    if to_hoist.metadata.is_none() {
        to_hoist.metadata = kube_schema.metadata.take();
    }
```

clux commented on kube-core/src/schema/transform_any_of.rs:
```diff
@@ -0,0 +1,238 @@
+
+    #[test]
+    fn optional_tagged_enum_with_unit_variants_but_also_an_existing_description() {
+        let original_schema_object_value = serde_json::json!({
```
i think these tests are nice.

minor nit; variables length here causes my eyes to rescan quite a bit. we can remove the substring `_schema_object` from all the variables here to make it slightly less verbose as all variables have this substring.

clux commented on kube-core/src/schema/transform_one_of.rs:
```diff
@@ -0,0 +1,181 @@
+            ..
+        }) => r#type,
+        Schema::Object(untyped) => panic!("oneOf variants need to define a type: {untyped:#?}"),
+        Schema::Bool(_) => panic!("oneOf variants can not be of type boolean"),
```
is this a golang restriction or something related to how we have a weird `Schema` enum with a  "trivial boolean JSON schema" embedded? (which i tbh do not remember why we have)

clux commented on kube-core/src/schema/transform_one_of.rs:
```diff
@@ -0,0 +1,181 @@
+    // Ensure all variant types match it
+    if types.any(|r#type| r#type != variant_type) {
+        panic!("oneOf variants must all have the same type");
+    }
```
probably a sane unwrap for now. might need another line about why this is true in the code.

NickLarsenNZ replied:
Added extra explanation in 77af8e5d4dec04daa6ad904990edbd7b8c7b5f4b

clux commented on kube-derive/tests/crd_schema_test.rs:
```diff
@@ -49,6 +49,7 @@ struct FooSpec {
     #[serde(skip_serializing_if = "Option::is_none")]
     nullable_skipped: Option<String>,
     nullable: Option<String>,
+    nullnullnullable: Option<Option<Option<String>>>,
```
variable names to the tune of `row row row yourboat`

clux reviewed (commented):
some initial comments on various bits. suggestions on some code-minimisation.
feels like i mostly want more use of slice-patterns (to remove some implicit panics), more let-else use, more comments on the invariants that cause us to panic, and possibly (possibly also testing of the interface direct as per previous comment).

otherwise, i think this approach is nice. it is very thorough, and you're dealing with a more complicated codebase at stackable than i have to deal with so happy to take changes like this from someone with a bit of skin in the game.

Danil-Grigorev commented on kube-core/src/schema/transform_any_of.rs:
```diff
@@ -0,0 +1,238 @@
+///
+/// NOTE: This should work regardless of whether other hoisting has been performed or not.
+#[derive(Debug, Clone)]
+pub struct AnyOfSchemaRewriter;
```
I believe that this method has no functional differences with implementation in https://github.com/kube-rs/kube/pull/1853 and therefore will pass the tests. We just try to panic more, where we should not.

Danil-Grigorev commented on kube-core/src/schema/transform_any_of.rs:
```diff
@@ -0,0 +1,238 @@
+    // TODO (@NickLarsenNZ): Return errors instead of panicking, leave panicking up to the
+    // infallible schemars::Transform
+    let Schema::Object(to_hoist) = subschema_to_hoist else {
+        panic!("the non-null anyOf subschema is a bool. That is not expected here");
```
We should not panic in this code. This produces no linter hints, and gives no trace on which field actually failed. The only correct approach with the chosen path by `schemars` is to split code into more domain specific parts, and log errors instead of panicking here.

Danil-Grigorev reviewed (commented):
> I guess I'd prefer not to, partly because of the possible merge conflicts, but also because I think they only solve part of the problem without seeing the broader issue.

I believe the https://github.com/kube-rs/kube/issues/1821 is only mentioning the specific issue with optional enums, and we need to fix that one directly, and not "oh btw I also fixed that". Having separate issues for this PR addresses would help.

NickLarsenNZ committed:
chore(kube-core): use let-else

NickLarsenNZ committed:
chore(kube-core): simplify metadata handling

NOTE: the kube_schema metadata does need dropping, but that happens implicitly when we replace kube_schema with the subschema

NickLarsenNZ committed:
test(kube-core): Make variable names shorter

NickLarsenNZ committed:
chore(kube-core): Add further explanation about why the types must be the same

NickLarsenNZ commented on kube-core/src/schema/transform_one_of.rs:
```diff
@@ -0,0 +1,181 @@
+                ..
+            }) => r#enum.clone(),
+            Schema::Object(SchemaObject { other, .. }) => other.get("const").cloned().into_iter().collect(),
+            Schema::Bool(_) => panic!("oneOf variants can not be of type boolean"),
```
We could do it, but I'm of the opinion that leaving it as is prevents strange behavior should code changes further up be made.

I'm happy to change it to `_ => None` if you would like, though.

There might be a way to combine the two iterators, and maybe it is easier with the `ReplaceConstValue` transformer (I'd prefer to leave that as a future improvement, if possible).

NickLarsenNZ commented on kube-core/src/schema/transform_properties.rs:
```diff
@@ -0,0 +1,806 @@
+
+    // Ensure we aren't looking at the subschema with a null, as that is hoisted by another
+    // transformer.
+    if subschemas.len() == 2 {
```
I quickly tried both:

```rust
match subschemas.len() { ...}
// and
match subschemas.as_slice() { .. }
```

But it didn't make things any simpler.

Would you mind giving a rough example as a hint?

NickLarsenNZ committed:
dco signoff

Signed-off-by: Nick Larsen <nick.larsen@stackable.tech>

clux commented:
Hey. I have had a bit more of a look at this case, and just wanted to write a bit here because this is inherently slow pr as is, and I while I appreciate the effort that's gone in here, i also want emphasise that it's not an easy pr to expedite.

This PR does a lot of schema modifications. Most of them do indeed look sensible on the surface, but I/we don't have a complete picture of all it can impact, and because there are so many, i am afraid to blanket approve or even guide them given the scope and potential impact. With potentially controversial solutions with regard to how things are split, what to handle, how to handle errors etc, this is even harder. Like, I can make pragmatic guesses (as i have done in the comments here) as i do for my own personal stuff w.r.t. panics etc, but it's harder to justify in a generic library with tons of users.

So, this makes individual solutions more attractive. Have merged Danil's PR in #1853 initially because that at least solves the main thing we need to have fixed before 3.0/kubernetes1.35, and it unblocks me fixing some other things.

I am not saying we don't want the remaining pieces in here, but please try to make it easy for us to accept them. If you are able please try to;

- raise issues about the specific lacks (after #1853) that you would like solved. it's nice to have some discussion about why/what, future-proofing is needed, before looking at an implementation without all the context. The option/nullable concern clearly needed fixing, and we could prove that with the attempts and reactions on https://github.com/kube-rs/kube/issues/1821
- split the separate remaining features into separate prs (so that not every controversial decision/uncertainty has to be decided at the same time)

because not doing this leads to uncertainty, and uncertainty with volunteer maintainers is generally a recipe for them thinking;  "eehh, i'll look at this tomorrow" (indefinitely). I am sorry.

(If you decide that this is too much hassle and would like to keep some part of the schema rewriting in-house on your end, that's obviously an option, just know that we do actually want to solve the 99% cases in here.)

NickLarsenNZ commented:
Hi @clux, apologies for the delayed reply. I'm now back after falling ill last week.

> I am not saying we don't want the remaining pieces in here, but please try to make it easy for us to accept them. If you are able please try to;
> - ...
> - split the separate remaining features into separate prs (so that not every controversial decision/uncertainty has to be decided at the same time)

I can give that a try, although the tests would then fail until all things are fixed.

I did have a chat with @sbernauer, and he suggested raising a PR with only the new tests, and then:
- Marking failing tests as skipped, and linking to a new issue
- Raising separate PRs (like you suggested) to resolve each issue, while keeping the changeset small.

Would you be happy with that approach?

NickLarsenNZ commented on kube-core/src/schema/transform_any_of.rs:
```diff
@@ -0,0 +1,238 @@
+    use super::*;
+
+    #[test]
+    fn optional_tagged_enum_with_unit_variants() {
```
This was a difficult problem to solve with the previous code. So I made a point to add commentary in case of future changes to generated schemars schemas.

I made a point to make these unit tests to show exactly what the transformation function is intended to operate on. That way, if a regression creeps in, the schemars output can be checked, the unit test updated, and the transformation function fixed (ideally the integration test would be unchanged).

I can of course remove the tests, but they are intentional.

NickLarsenNZ commented on kube-core/src/schema/transform_optional_enum_with_null.rs:
```diff
@@ -0,0 +1,101 @@
+/// NOTE: The trailing null is removed because it's not needed by Kubernetes
+/// and makes the CRD more compact by removing redundant information.
+#[derive(Debug, Clone)]
+pub struct OptionalEnumSchemaRewriter;
```
Per previous comment, will look at combining them if they relate well enough.

> Why is this needed? Can this be produced/tested by some CRD schema?

I not quite sure what you're asking

clux commented on kube-derive/tests/crd_mixed_enum_test.rs:
```diff
@@ -0,0 +1,396 @@
+fn valid_enum_3() {
+    assert_json_eq!(
+        ValidEnum3::crd(),
+        json!(
```
you might be able to make these integration tests a little less intimidating by only testing the main version output;


```suggestion
        ValidEnum3::crd().spec.versions[0],
        json!( ...skip first 16 lines
```
